File: #20_isValid.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '{' or s[i] == '[':
                stack.append(s[i])
            if len(s) == 0:
                return False
            if s[i] == ')':
                if len(stack) == 0:
                    return False
                else:
                    x = stack.pop()
                    if x == '{' or x == '[':
                        return False
            elif s[i] == '}':
                if len(stack) == 0:
                    return False
                else:
                    x = stack.pop()
                    if x == '(' or x == '[':
                        return False
            elif s[i] == ']':
                if len(stack) == 0:
                    return False
                else:
                    x = stack.pop()
                    if x == '{' or x == '(':
                        return False
        if len(stack) == 0:
            return True
        else:
            return False

File: Java/#20_isValid.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> s1 = new Stack<Character>();
        for(int i=0;i<s.length();i++){
            char c= s.charAt(i);
            if(c=='(' || c=='[' || c=='{') s1.push(c);
            else{
                if(!s1.isEmpty()){
                char open=s1.pop();
                if(c==')' && open=='(') continue;
                else if(c==']' && open=='[') continue;
                else if(c=='}' && open=='{') continue;
                else return false;
                }else return false;
            }
        }
        return s1.isEmpty();
    }
}
